File: WebKitTools/Scripts/webkitpy/layout_tests/port/http_server.py
#!/usr/bin/env python
# Copyright (C) 2010 Google Inc. All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
#
#     * Redistributions of source code must retain the above copyright
# notice, this list of conditions and the following disclaimer.
#     * Redistributions in binary form must reproduce the above
# copyright notice, this list of conditions and the following disclaimer
# in the documentation and/or other materials provided with the
# distribution.
#     * Neither the name of Google Inc. nor the names of its
# contributors may be used to endorse or promote products derived from
# this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

"""A class to help start/stop the lighttpd server used by layout tests."""

from __future__ import with_statement

import codecs
import logging
import optparse
import os
import shutil
import subprocess
import sys
import tempfile
import time
import urllib

import factory
import http_server_base

_log = logging.getLogger("webkitpy.layout_tests.port.http_server")


class HttpdNotStarted(Exception):
    pass


class Lighttpd(http_server_base.HttpServerBase):

    def __init__(self, port_obj, output_dir, background=False, port=None,
                 root=None, run_background=None):
        """Args:
          output_dir: the absolute path to the layout test result directory
        """
        # Webkit tests
        http_server_base.HttpServerBase.__init__(self, port_obj)
        self._output_dir = output_dir
        self._process = None
        self._port = port
        self._root = root
        self._run_background = run_background
        if self._port:
            self._port = int(self._port)

        try:
            self._webkit_tests = os.path.join(
                self._port_obj.layout_tests_dir(), 'http', 'tests')
            self._js_test_resource = os.path.join(
                self._port_obj.layout_tests_dir(), 'fast', 'js', 'resources')
        except:
            self._webkit_tests = None
            self._js_test_resource = None

        # Self generated certificate for SSL server (for client cert get
        # <base-path>\chrome\test\data\ssl\certs\root_ca_cert.crt)
        self._pem_file = os.path.join(
            os.path.dirname(os.path.abspath(__file__)), 'httpd2.pem')

        # One mapping where we can get to everything
        self.VIRTUALCONFIG = []

        if self._webkit_tests:
            self.VIRTUALCONFIG.extend(
               # Three mappings (one with SSL) for LayoutTests http tests
               [{'port': 8000, 'docroot': self._webkit_tests},
                {'port': 8080, 'docroot': self._webkit_tests},
                {'port': 8443, 'docroot': self._webkit_tests,
                 'sslcert': self._pem_file}])

    def is_running(self):
        return self._process != None

    def start(self):
        if self.is_running():
            raise 'Lighttpd already running'

        base_conf_file = self._port_obj.path_from_webkit_base('WebKitTools',
            'Scripts', 'webkitpy', 'layout_tests', 'port', 'lighttpd.conf')
        out_conf_file = os.path.join(self._output_dir, 'lighttpd.conf')
        time_str = time.strftime("%d%b%Y-%H%M%S")
        access_file_name = "access.log-" + time_str + ".txt"
        access_log = os.path.join(self._output_dir, access_file_name)
        log_file_name = "error.log-" + time_str + ".txt"
        error_log = os.path.join(self._output_dir, log_file_name)

        # Remove old log files. We only need to keep the last ones.
        self.remove_log_files(self._output_dir, "access.log-")
        self.remove_log_files(self._output_dir, "error.log-")

        # Write out the config
        with codecs.open(base_conf_file, "r", "utf-8") as file:
            base_conf = file.read()

        # FIXME: This should be re-worked so that this block can
        # use with open() instead of a manual file.close() call.
        # lighttpd.conf files seem to be UTF-8 without BOM:
        # http://redmine.lighttpd.net/issues/992
        f = codecs.open(out_conf_file, "w", "utf-8")
        f.write(base_conf)

        # Write out our cgi handlers.  Run perl through env so that it
        # processes the #! line and runs perl with the proper command
        # line arguments. Emulate apache's mod_asis with a cat cgi handler.
        f.write(('cgi.assign = ( ".cgi"  => "/usr/bin/env",\n'
                 '               ".pl"   => "/usr/bin/env",\n'
                 '               ".asis" => "/bin/cat",\n'
                 '               ".php"  => "%s" )\n\n') %
                                     self._port_obj._path_to_lighttpd_php())

        # Setup log files
        f.write(('server.errorlog = "%s"\n'
                 'accesslog.filename = "%s"\n\n') % (error_log, access_log))

        # Setup upload folders. Upload folder is to hold temporary upload files
        # and also POST data. This is used to support XHR layout tests that
        # does POST.
        f.write(('server.upload-dirs = ( "%s" )\n\n') % (self._output_dir))

        # Setup a link to where the js test templates are stored
        f.write(('alias.url = ( "/js-test-resources" => "%s" )\n\n') %
                    (self._js_test_resource))

        # dump out of virtual host config at the bottom.
        if self._root:
            if self._port:
                # Have both port and root dir.
                mappings = [{'port': self._port, 'docroot': self._root}]
            else:
                # Have only a root dir - set the ports as for LayoutTests.
                # This is used in ui_tests to run http tests against a browser.

                # default set of ports as for LayoutTests but with a
                # specified root.
                mappings = [{'port': 8000, 'docroot': self._root},
                            {'port': 8080, 'docroot': self._root},
                            {'port': 8443, 'docroot': self._root,
                             'sslcert': self._pem_file}]
        else:
            mappings = self.VIRTUALCONFIG
        for mapping in mappings:
            ssl_setup = ''
            if 'sslcert' in mapping:
                ssl_setup = ('  ssl.engine = "enable"\n'
                             '  ssl.pemfile = "%s"\n' % mapping['sslcert'])

            f.write(('$SERVER["socket"] == "127.0.0.1:%d" {\n'
                     '  server.document-root = "%s"\n' +
                     ssl_setup +
                     '}\n\n') % (mapping['port'], mapping['docroot']))
        f.close()

        executable = self._port_obj._path_to_lighttpd()
        module_path = self._port_obj._path_to_lighttpd_modules()
        start_cmd = [executable,
                     # Newly written config file
                     '-f', os.path.join(self._output_dir, 'lighttpd.conf'),
                     # Where it can find its module dynamic libraries
                     '-m', module_path]

        if not self._run_background:
            start_cmd.append(# Don't background
                             '-D')

        # Copy liblightcomp.dylib to /tmp/lighttpd/lib to work around the
        # bug that mod_alias.so loads it from the hard coded path.
        if sys.platform == 'darwin':
            tmp_module_path = '/tmp/lighttpd/lib'
            if not os.path.exists(tmp_module_path):
                os.makedirs(tmp_module_path)
            lib_file = 'liblightcomp.dylib'
            shutil.copyfile(os.path.join(module_path, lib_file),
                            os.path.join(tmp_module_path, lib_file))

        env = self._port_obj.setup_environ_for_server()
        _log.debug('Starting http server')
        # FIXME: Should use Executive.run_command
        self._process = subprocess.Popen(start_cmd, env=env)

        # Wait for server to start.
        self.mappings = mappings
        server_started = self.wait_for_action(
            self.is_server_running_on_all_ports)

        # Our process terminated already
        if not server_started or self._process.returncode != None:
            raise google.httpd_utils.HttpdNotStarted('Failed to start httpd.')

        _log.debug("Server successfully started")

    # TODO(deanm): Find a nicer way to shutdown cleanly.  Our log files are
    # probably not being flushed, etc... why doesn't our python have os.kill ?

    def stop(self, force=False):
        if not force and not self.is_running():
            return

        httpd_pid = None
        if self._process:
            httpd_pid = self._process.pid
        self._port_obj._shut_down_http_server(httpd_pid)

        if self._process:
            # wait() is not threadsafe and can throw OSError due to:
            # http://bugs.python.org/issue1731717
            self._process.wait()
            self._process = None

File: Tools/TestWebKitAPI/Helpers/cocoa/HTTPServer.swift
// Copyright (C) 2025 Apple Inc. All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions
// are met:
// 1. Redistributions of source code must retain the above copyright
//    notice, this list of conditions and the following disclaimer.
// 2. Redistributions in binary form must reproduce the above copyright
//    notice, this list of conditions and the following disclaimer in the
//    documentation and/or other materials provided with the distribution.
//
// THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
// THE POSSIBILITY OF SUCH DAMAGE.

#if ENABLE_CXX_INTEROP

import Foundation
private import TestWebKitAPILibrary.Helpers.cocoa.HTTPServer
import struct Swift.String

/// A description of an HTTP server route with a path and a response.
///
/// Typically, a simple route can be created using a path and some response data:
///
/// ```swift
/// Route("/hello") {
///     "Hello World!"
/// }
/// ```
///
/// A route can also be used to express a group of routes that share a common path prefix:
///
/// ```swift
/// Route("/parent") {
///     Route("/a") {
///         "My path is /parent/a"
///     }
///
///     Route("/b") {
///         "My path is /parent/b"
///     }
/// }
/// ```
public struct Route: Sendable {
    fileprivate struct Storage: Sendable {
        let pathComponents: [String]
        let response: String
    }

    fileprivate let children: [Storage]

    fileprivate init(children: [Storage]) {
        self.children = children
    }

    fileprivate init(path: String, response: String) {
        self.children = [Storage(pathComponents: [path], response: response)]
    }

    /// Creates a Route from a group of child Routes.
    ///
    /// - Parameters:
    ///   - path: The path of this route. If this value is non-empty, it must start with `/`.
    ///   - route: The children of this route; each child has it's full path prepended by the path of this route.
    public init(_ path: String, @RouteBuilder _ route: () -> Route) {
        self.children = route().children
            .map {
                Storage(pathComponents: [path] + $0.pathComponents, response: $0.response)
            }
    }

    /// Creates a Route whose response data is a String.
    ///
    /// - Parameters:
    ///   - path: The path of this route. If this value is non-empty, it must start with `/`.
    ///   - response: The response to be used.
    public init(_ path: String, _ response: () -> String) {
        self.init(path: path, response: response())
    }
}

/// A result builder used to create a ``Route``.
@resultBuilder
public struct RouteBuilder {
    /// Create a ``Route`` from a group of Routes.
    public static func buildBlock(_ components: Route...) -> Route {
        .init(children: components.flatMap(\.children))
    }
}

/// A type used to simulate an HTTP server with a set of predefined responses for different types of requests.
@MainActor
@safe
public struct HTTPServer: ~Copyable {
    /// A protocol describing how an HTTP connection handles requests.
    public enum `Protocol`: Sendable {
        /// The HTTP protocol.
        case http

        /// The HTTPS protocol.
        case https

        /// The HTTPS protocol, using a legacy version of TLS.
        case httpsWithLegacyTLS

        /// The HTTP2 protocol.
        case http2

        /// The HTTPS proxy protocol.
        case httpsProxy

        /// The HTTPS proxy protocol with authentication.
        case httpsProxyWithAuthentication
    }

    private var storage: TestWebKitAPI.__CxxHTTPServer

    /// Create a server from a group of routes.
    ///
    /// For example, a server which loads `/webkit/success` after loading `/example` can be created using two routes:
    ///
    /// ```swift
    /// func doSomethingCool() async throws {
    ///     let server = HTTPServer(protocol: .httpsProxy) {
    ///         Route("/example") {
    ///             "<script>w = window.open('https://webkit.org/webkit/success')</script>"
    ///         }
    ///
    ///         Route("/webkit/success") {
    ///             "Loaded!"
    ///         }
    ///     }
    ///
    ///     let page = WebPage()
    ///
    ///     let result = try await server.run {
    ///         let destination = address.appendingPathComponent("example")
    ///
    ///         try await page.load(destination).wait()
    ///
    ///         return page.url
    ///     }
    ///
    ///     // `result` is now a `URL` with a value like `http://127.0.0.1:8080/webkit/success`.
    /// }
    /// ```
    ///
    /// - Parameters:
    ///   - protocol: The HTTP protocol to use for this server.
    ///   - route: A group of routes that correspond to a mapping of request paths to responses.
    public init(protocol: `Protocol`, @RouteBuilder _ route: () -> Route) {
        var entries = unsafe TestWebKitAPI.__CxxHTTPServer.ResponseMap()

        let routes = route().children
        for child in routes {
            let path = child.pathComponents.joined()
            let response = unsafe TestWebKitAPI.HTTPResponse(WTF.String(child.response))

            unsafe hashMapSet(&entries, consuming: .init(path), consuming: response)
        }

        unsafe self.storage = .init(consuming: entries, .init(`protocol`), consuming: .init(), nil, .init(), .Yes)
    }

    /// Calls the given closure after starting the server, and then closes the server once finished.
    ///
    /// - Parameter body: A closure that will run while this server is active.
    /// - Returns: The return value, if any, of the `body` closure parameter.
    /// - Throws: Any error thrown by `body`.
    public mutating func run<Result, E>(
        _ body: () async throws(E) -> sending Result
    ) async throws(E) -> sending Result where E: Error, Result: ~Copyable {
        await withCheckedContinuation { continuation in
            unsafe self.storage.startListening(
                consuming: .init(
                    {
                        continuation.resume()
                    },
                    WTF.ThreadLikeAssertion(WTF.CurrentThreadLike())
                )
            )
        }

        let result = try await body()

        await withCheckedContinuation { continuation in
            unsafe self.storage.cancel(
                consuming: .init(
                    {
                        continuation.resume()
                    },
                    WTF.ThreadLikeAssertion(WTF.CurrentThreadLike())
                )
            )
        }

        return result
    }
}

extension TestWebKitAPI.__CxxHTTPServer.`Protocol` {
    fileprivate init(_ protocol: HTTPServer.`Protocol`) {
        self =
            switch `protocol` {
            case .http: .Http
            case .https: .Https
            case .httpsWithLegacyTLS: .HttpsWithLegacyTLS
            case .http2: .Http2
            case .httpsProxy: .HttpsProxy
            case .httpsProxyWithAuthentication: .HttpsProxyWithAuthentication
            }
    }
}

#endif // ENABLE_CXX_INTEROP
